### PR TITLE
WRF: Add version 3.9.1.1 and improve recipe robustness

### PR DESCRIPTION
**Pull request by ptooley**
PR to improve several aspects of WRF support in Spack:

The headline issue here is the fact that upstream use `make -i -k` and their custom make script always returns 0. This means that the standard approach of checking the return code won't detect a failed build.  Instead we need to actually check the output of the script to see if it succeeded.  My approach is the least-bad solution I could find.

Second big issue is that the configure script *enforces* interactivity and dynamically generates its option.  Previous recipe version had response values hardcoded for an assumed x86 platform.  This new version should be platform independent and has been tested on at least aarch64 and x86_64.

Additionally I have added support for v 3.9 which is a common HPC benchmarking workload, and fixed an issue with failure to compile against recent glibcs that don't ship sunrpc headers.

* Include version 3.9.1.1 as common benchmarking workload
* Fix compilation against recent glibc (use spack installed libtirpc)
* Detect and handle failed compilation (upstream use make -i)

### CODE REVIEW

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/wrf/package.py]: Can you add your name to the list of maintainers instead of replacing @MichaelLaufer?

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/wrf/package.py]:
```suggestion
    depends_on("pkgconfig", type=("build"))
```
Spack has a virtual package that lets you use `pkg-config` or `pkgconf`

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/wrf/package.py]: I assume this is the level of parallelism? In that case:
```suggestion
                str(make_jobs),
```

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/wrf/package.py]: Instead of using `Popen`, is it possible to use Spack's `Executable` class? Something like:
```python
csh = Executable(csh_bin)
csh(...)
```

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/wrf/package.py]:
```suggestion
        csh_bin = self.spec["tcsh"].prefix.bin.csh
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]
@MichaelLaufer can you review this?

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (ptooley) [reply var/spack/repos/builtin/packages/wrf/package.py]: Yes, sorry. This started life as a new recipe because the changes were so big.

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (ptooley) [reply var/spack/repos/builtin/packages/wrf/package.py]: This is result of experiments where I was still getting seemingly random compilation failures with more than this many build jobs.  The previous recipe was using up to 20 but that was still not working well for me.  I'll reinstate the up-to-twenty logic for now...

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (ptooley) [reply var/spack/repos/builtin/packages/wrf/package.py]: Having looked at the source for the Executable class I think I can, but the text output from that compile stage is massive, so I was trying to avoid keeping the whole build log in RAM.  If spack is doing that anyway behind the scenes then I think I can convert this one to `Executable` but not the interactive configure step.

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/wrf/package.py]: You can also use up-to-10 if that helps. Did I also see you use `parallel=False`? I think that would cap at 1.

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/wrf/package.py]: Ah, I don't think we're doing that, didn't realize this package needed to be so complicated.

Comment (ptooley) [reply var/spack/repos/builtin/packages/wrf/package.py]: Testing with `Executable` gives me the behaviour I need, thanks!

**Comment by ptooley** [comment]
All changes included and I also cleaned up some of the more egregious formatting choices my autoformatter made...

Apologies for the force push - I tried to rebase my local repo onto the latest develop and got in a complete mess.

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/wrf/package.py:187]: `self.compiler.cc` and friends are the actual compilers, generally we want to use Spack's compiler wrappers (`spack_cc`) so that we automatically inject the appropriate linking args. Does it not work if you use Spack's compiler wrappers?

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/wrf/package.py]:
```suggestion
        if self.spec.satisfies("@3.9.1.1 %gcc"):
```

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/wrf/package.py]: We could use `self.spec['mpi'].mpicc` and friends here instead, that may be more robust as not every MPI library uses the same names.

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/wrf/package.py]: I think the old logic of `str(min(int(make_jobs), 20))` is more clear.

**Review by adamjstewart — COMMENTED** [review]
Few more comments on the Spack recipe. I don't know much about WRF so I can't review the patches.

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (ptooley) [reply var/spack/repos/builtin/packages/wrf/package.py:187]: Here I am picking up the actual name of the compiler binary to compare with the options presented by the interactive configure script.  It doesn't actually inject any paths.

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (ptooley) [reply var/spack/repos/builtin/packages/wrf/package.py]: Done

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (ptooley) [reply var/spack/repos/builtin/packages/wrf/package.py]: Done

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (ptooley) [reply var/spack/repos/builtin/packages/wrf/package.py]: Done

**Review by adamjstewart — APPROVED** [review]
Looks good to me from my end. Since this is a pretty big change I'm still hoping for a review from @MichaelLaufer or another WRF user. Let's give it a few days. If we don't hear back, ping me and I'll merge.

File: var/spack/repos/builtin/packages/wrf/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/wrf/package.py]: The unit tests are failing on this line. I don't think you can open two file handles on the same line. You should doubly nest this.

Comment (ptooley) [reply var/spack/repos/builtin/packages/wrf/package.py]: Yeah Python 2.6 chokes on this, 2.7 onwards it is okay. Vermin didn't flag it and I can't even install Python2.6 from my package manager to test.

**Comment by ptooley** [comment]
> Looks good to me from my end. Since this is a pretty big change I'm still hoping for a review from @MichaelLaufer or another WRF user. Let's give it a few days. If we don't hear back, ping me and I'll merge.

Thanks! Do you want this squashed to a single commit before merge?

**Comment by adamjstewart** [comment]
> Do you want this squashed to a single commit before merge?

Nah, we automatically squash and merge all PRs, so there's no need to do it yourself.

**Comment by MichaelLaufer** [comment]
> 
> 
> Looks good to me from my end. Since this is a pretty big change I'm still hoping for a review from @MichaelLaufer or another WRF user. Let's give it a few days. If we don't hear back, ping me and I'll merge.

@ptooley @adamjstewart 
We are having some IT issues on our cluster at the moment so I could not test this out. I hope to have it fixed by tomorrow. Looks good, and I am happy to see V3.9.1.1 support!

**Comment by adamjstewart** [comment]
In that case I'll merge for now and we can backport any changes if they are needed later. Thanks for reviewing @MichaelLaufer!

**Pull request merged**
